doc/services/rtio/index.rst

```diff
@@ -0,0 +1,417 @@
+hardware knowledge or leaky abstractions over DMA controllers. Neither is ideal.
+
+To enable asynchronous operations, especially with DMA, a description of what
+to do rather than direct operations through C and callbacks is needed. Enabling
+DMA features such as channels with priority, and sequences of transfers requires
+more than a simple list of descriptions.
+
+Using DMA and/or interrupt driven I/O shouldn't dictate whether or not the
+call is blocking or not.
+
+Inspiration, introducing io_uring
+*********************************
+
+It's better not to reinvent the wheel (or ring in this case) and io_uring as an
+API from the Linux kernel provides a winning model. In io_uring there are two
+lock-free ring buffers acting as queues shared between the kernel and a userland
+application. One queue for submission entries which may be chained and flushed to
+create concurrent sequential requests. A second queue for completion events.
+Only a single syscall is actually required to execute many operations, the
+io_uring_submit call. This call may block the caller when a number of
+operations to wait on is given.
+
+This model maps well to DMA and interrupt driven transfers. A request to do a
+sequence of operations in an asynchronous way directly relates
+to the way hardware typically works with interrupt driven state machines
+potentially involving multiple peripheral IPs like bus and DMA controllers.
+
+Submission Queue and Chaining
+*****************************
+
+The submission queue (sq), is the description of the operations
+to perform in concurrent chains.
+
+For example imagine a typical SPI transfer where you wish to write a
+register address to then read from. So the sequence of operations might be...
+
+   1. Chip Select
+   2. Clock Enable
+   3. Write register address into SPI transmit register
+   4. Read from the SPI receive register into a buffer
+   5. Disable clock
+   6. Disable Chip Select
+
+If anything in this chain of operations fails give up. Some of those operations
+can be embodied in a device abstraction that understand a read or write
+implicitly means setup the clock and chip select. The transactional nature of
+the request also needs to be embodied in some manner. Of the operations above
+perhaps the read could be done using DMA as its large enough make sense. That
+requires an understanding of how to setup the device's particular DMA to do so.
+
+The above sequence of operations is embodied in RTIO as chain of
+submission queue entries (sqe). Chaining is done by setting a bitflag in
+an sqe to signify the next sqe must wait on the current one.
+
+Because the chip select and clocking is common to a particular SPI controller
+and device on the bus it is embodied in what RTIO calls an iodev.
+
+Multiple operations against the same iodev are done in the order provided as
+soon as possible. If two operation chains have varying points using the same
+device its possible one chain will have to wait for another to complete.
```

Comment:
Consider wanting to read data from two devices on the same bus, and I believe the answer to this is potentially yes. That has some implications that I've been mulling over.

Reply:
For some higher level devices, I can understand a yes on this... If you have something like an I2C that one chain is doing a START/WRITE/READ/READ/RESTART/WRITE/STOP. That would not really map with the restart but otherwise would be a Connect/TX/RX/RX/????/TX/Disconnect. If it were required that this be synchronous each phase of this would happen one step at a time. If there were also another chain that was good to go that were talking to that same device it seems like this could interfer with the current thread that sort of owns the device until Disconnect...
Also the Disconnect would be locked out if there were a failure on one of the previous operations, say a timeout. Would that possibly lock out other chains or at least leave the bus in an odd state?

Reply:
Yeah this is all correct, the implications are that other chains would need to, as part of the request, enqueue themselves to be notified when the iodev is available. Maybe this can all be implied in a connect type request, where if chain 1 is connected and chain 2 then requests to connect, that chain 2 connect sqe won't complete until chain 1 fails or disconnects. A failure on a chain should implicitly disconnect in some manner, probably by calling an iodev provided API like chain_failed(), and the iodev can decide what to do if its in a transactional state.